metals/src/main/scala/scala/meta/internal/metals/MetalsLspWorkspaceFolderService.scala

```diff
@@ -0,0 +1,2764 @@
+import scala.meta.metals.lsp.ScalaLspService
+import scala.meta.parsers.ParseException
+import scala.meta.pc.CancelToken
+import scala.meta.pc.DisplayableException
+import scala.meta.tokenizers.TokenizeException
+
+import ch.epfl.scala.bsp4j.CompileReport
+import ch.epfl.scala.{bsp4j => b}
+import com.google.gson.JsonElement
+import com.google.gson.JsonNull
+import com.google.gson.JsonObject
+import com.google.gson.JsonPrimitive
+import io.undertow.server.HttpServerExchange
+import org.eclipse.lsp4j.ExecuteCommandParams
+import org.eclipse.lsp4j._
+import org.eclipse.lsp4j.jsonrpc.messages.{Either => JEither}
+import org.eclipse.{lsp4j => l}
+
+/**
+ * Metals implementation of the Scala Language Service.
+ * @param ec
+ *  Execution context used for submitting tasks. This class DO NOT manage the
+ *  lifecycle of this execution context.
+ * @param sh
+ *  Scheduled executor service used for scheduling tasks. This class DO NOT
+ *  manage the lifecycle of this executor.
+ * @param serverInputs
+ *  Collection of different parameters used by Metals for running,
+ *  which main purpose is allowing for custom bahaviour in tests.
+ * @param workspace
+ *  An absolute path to the workscape.
+ * @param client
+ *  Metals client used for sending notifications to the client. This class DO
+ *  NOT manage the lifecycle of this client. It is the responsibility of the
+ *  caller to shut down the client.
+ * @param initializeParams
+ *  Initialization parameters send by the client in the initialize request,
+ *  which is the first request sent to the server by the client.
+ */
+class MetalsLspWorkspaceFolderService(
+    ec: ExecutionContextExecutorService,
+    sh: ScheduledExecutorService,
+    serverInputs: MetalsServerInputs,
+    client: MetalsLanguageClient,
+    initializeParams: InitializeParams,
+    clientConfig: ClientConfiguration,
+    val workspace: AbsolutePath,
+    val workspaceId: String,
+) extends Cancelable
+    // TODO:: this shouldn't extend WorkspaceService -- those should be global commands
+    with ScalaLspService {
+  import serverInputs._
+
+  ThreadPools.discardRejectedRunnables("MetalsLanguageServer.sh", sh)
+  ThreadPools.discardRejectedRunnables("MetalsLanguageServer.ec", ec)
+
+  private val cancelables = new MutableCancelable()
+  val isCancelled = new AtomicBoolean(false)
+
+  override def cancel(): Unit = {
+    if (isCancelled.compareAndSet(false, true)) {
+      val buildShutdown = bspSession match {
+        case Some(session) => session.shutdown()
+        case None => Future.successful(())
+      }
+      try {
+        cancelables.cancel()
+      } catch {
+        case NonFatal(_) =>
+      }
+      try buildShutdown.asJava.get(100, TimeUnit.MILLISECONDS)
+      catch {
+        case _: TimeoutException =>
+      }
+    }
+  }
+
+  private implicit val executionContext: ExecutionContextExecutorService = ec
+
+  @volatile
+  private var userConfig: UserConfiguration = initialUserConfig
```

Comment:
This should be global for all workspaces I think.

Reply:
But you can have settings per workspace folder. Currently in metals vs code extension the only setting that can be set per folder is `Java Home`. 

_In the current version of the language server protocol, the client pushes settings to the server. With the introduction of a resource scope, this is not possible anymore since the actual settings values can depend on a resource. We introduce a protocol which allows servers to fetch settings from a client comparable to the getConfiguration API in the extension host. The protocol addition is documented [here](https://github.com/Microsoft/vscode-languageserver-node/blob/master/protocol/src/protocol.configuration.ts). The bundled CSS or HTML language extensions illustrate this approach._

Reply:
Hmm... I don't think `Java Home` should be set per folder. Metals can inly run using one Java version and we would run into a lot of issues if we allow each folder to use a different one.

Reply:
I'm pretty sure that's by accident. It's something to be fixed in the extension. This doesn't change the fact that some settings could be per folder if we wanted it that way. For things like show implicits this could make sense. Anyway, let's make user config this global in the first version.